api/src/main/java/bio/terra/tanagra/underlay/datapointer/BigQueryDataset.java

```diff
@@ -0,0 +1,154 @@
+package bio.terra.tanagra.underlay.datapointer;
+
+import bio.terra.tanagra.exception.SystemException;
+import bio.terra.tanagra.query.QueryExecutor;
+import bio.terra.tanagra.query.bigquery.BigQueryExecutor;
+import bio.terra.tanagra.serialization.datapointer.UFBigQueryDataset;
+import bio.terra.tanagra.underlay.DataPointer;
+import bio.terra.tanagra.underlay.FieldPointer;
+import bio.terra.tanagra.underlay.Literal;
+import bio.terra.tanagra.utils.GoogleBigQuery;
+import com.google.auth.oauth2.ServiceAccountCredentials;
+import com.google.cloud.bigquery.Field;
+import com.google.cloud.bigquery.LegacySQLTypeName;
+import com.google.cloud.bigquery.Schema;
+import com.google.common.collect.ImmutableMap;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Map;
+import org.apache.commons.text.StringSubstitutor;
+
+public final class BigQueryDataset extends DataPointer {
+  private final String projectId;
+  private final String datasetId;
+  private final Path serviceAccountKeyFile;
+  private ServiceAccountCredentials serviceAccountCredentials;
+  private GoogleBigQuery bigQueryService;
+  private BigQueryExecutor queryExecutor;
+
+  private BigQueryDataset(
+      String name, String projectId, String datasetId, Path serviceAccountKeyFile) {
+    super(name);
+    this.projectId = projectId;
+    this.datasetId = datasetId;
+    this.serviceAccountKeyFile = serviceAccountKeyFile;
+  }
+
+  public static BigQueryDataset fromSerialized(UFBigQueryDataset serialized) {
+    if (serialized.getProjectId() == null || serialized.getProjectId().isEmpty()) {
+      throw new IllegalArgumentException("No BigQuery project ID defined");
+    }
+    if (serialized.getDatasetId() == null || serialized.getDatasetId().isEmpty()) {
+      throw new IllegalArgumentException("No BigQuery dataset ID defined");
+    }
+    if (serialized.getServiceAccountKeyFile() == null
+        || serialized.getServiceAccountKeyFile().isEmpty()) {
+      throw new IllegalArgumentException(
+          "No service account key file defined for BigQuery dataset");
+    }
+    Path serviceAccountKeyFile = Path.of(serialized.getServiceAccountKeyFile());
```

Comment:
I don't think this SA key file is going to be the long-term solution, mostly because GCP is recommending people move away from key files. Currently this SA key is used only for indexing. I did this just for expediency, particularly for running the tests.

I don't know whether the indexing/service SAs will typically be different or not. Indexing needs to read source data, write index data; service needs to read both. We should discuss requirements/user-friendly options for pointing Tanagra at credentials (e.g. do we need to handle >1 for a single underlay), maybe in a follow-on PR though.

Reply:
Maybe just a TODO for now then?